app/scripts/lib/extension-update-manager.ts

```diff
@@ -0,0 +1,77 @@
+/**
+ * Manages extension update detection and application.
+ * Provides a clean interface for detecting and applying extension updates
+ * during idle periods to minimize user disruption.
+ */
+
+import browser from 'webextension-polyfill';
+
+export class ExtensionUpdateManager {
+  private updatePending = false;
+
+  private isIdle = false;
+
+  /**
+   * Initializes the update manager by setting up event listeners
+   * for update detection.
+   */
+  public initialize(): void {
+    // Ensure we start in a non-idle state during initialization
+    this.isIdle = false;
+
+    browser.runtime.onUpdateAvailable.addListener(
+      this.handleUpdateAvailable.bind(this),
+    );
+  }
+
+  /**
+   * Sets the idle state of the extension.
+   *
+   * @param idle - Whether the extension is currently idle
+   */
+  public setIdleState(idle: boolean): void {
+    this.isIdle = idle;
+
+    // If we're now idle and there's a pending update, apply it immediately
+    if (idle && this.updatePending) {
+      this.applyPendingUpdateIfNeeded();
+    }
+  }
+
+  /**
+   * Checks if there's a pending update and applies it if needed.
+   * This should be called during idle periods when it's safe to reload
+   * the extension without disrupting user activity.
+   */
+  public applyPendingUpdateIfNeeded(): void {
+    if (this.updatePending && this.isIdle) {
+      try {
+        browser.runtime.reload();
+      } catch (error) {
+        console.error('Failed to reload extension:', error);
```

Comment:
Would we ever expect this to fail?

Reply:
I was thinking about browser failure or extension corrupt, but it will not be too much edge case to consider as we have comprehensive strategy for state corruption. Updated here 4a1a4f427891c56d77df6e84809698ca7da553d8